lightning/mydump/csv_parser.go

```diff
@@ -0,0 +1,151 @@
+package mydump
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb-lightning/lightning/config"
+	"github.com/pingcap/tidb-lightning/lightning/worker"
+)
+
+type CSVParser struct {
+	blockParser
+
+	rowBuf bytes.Buffer
+	cfg    *config.CSVConfig
+
+	delim1 []byte
+	delim2 []byte
+}
+
+func NewCSVParser(
+	cfg *config.CSVConfig,
+	reader io.Reader,
+	blockBufSize int64,
+	ioWorkers *worker.Pool,
+) *CSVParser {
+	return &CSVParser{
+		blockParser: makeBlockParser(reader, blockBufSize, ioWorkers),
+		cfg:         cfg,
+
+		delim1: []byte(cfg.Delimiter),
+		delim2: []byte(cfg.Delimiter + cfg.Delimiter),
+	}
+}
+
+type csvToken byte
+
+const (
+	csvTokNil csvToken = iota
+	csvTokSep
+	csvTokNewLine
+	csvTokField
+	csvTokQuotedField
+)
+
+func (parser *CSVParser) appendEmptyValues(sepCount int) {
+	var content string
+	if !parser.cfg.NotNull && parser.cfg.Null == "" {
+		content = ",NULL"
+	} else {
+		content = ",''"
+	}
+	for i := 1; i < sepCount; i++ {
+		parser.rowBuf.WriteString(content)
+	}
+}
+
+func (parser *CSVParser) appendField(content []byte, shouldUnquote bool, quote byte) {
+	if !parser.cfg.BackslashEscape {
+		content = bytes.Replace(content, []byte(`\`), []byte(`\\`), -1)
+	}
+	if shouldUnquote {
+		content = bytes.Replace(content[1:len(content)-1], parser.delim2, parser.delim1, -1)
+	}
+	if !parser.cfg.NotNull && parser.cfg.Null == string(content) {
+		parser.rowBuf.WriteString(",NULL")
+		return
+	}
+
+	parser.rowBuf.WriteByte(',')
+	parser.rowBuf.WriteByte(quote)
+	parser.rowBuf.Write(bytes.Replace(content, []byte{quote}, []byte{quote, quote}, -1))
+	parser.rowBuf.WriteByte(quote)
+}
+
+// ReadRow reads a row from the datafile.
+func (parser *CSVParser) ReadRow() error {
+	emptySepCount := 0
+	hasField := false
+
+	// skip the header first
+	if parser.pos == 0 && parser.cfg.Header {
+	outside:
+		for {
+			tok, content, err := parser.lex()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			switch tok {
+			case csvTokSep:
+			case csvTokField:
```

Comment:
Is it necessary to check the validity of the header? e.g: `col1,col2,,,col5`.

Reply:
We could; though it is probably unnecessary in this PR. With `col1,col2,,,col5`, this will produce either one of

```sql
INSERT INTO `t` (`col1`, `col2`, NULL, NULL, `col5`) VALUES ...;
INSERT INTO `t` (`col1`, `col2`, ``, ``, `col5`) VALUES ...;
```

and both will yield errors (syntax error or "unknown column" error)

Now, these *will* be a problem in #136 where we no longer assemble the SQL statement, but these functions will need to be rewritten anyway.